### PR TITLE
Implement `Clone` for more arrays

### PR DESCRIPTION
**Pull request by tbu-**
[breaking-change]

### CODE REVIEW

File: src/libcore/lib.rs [hunk 1]
```diff
@@ -85,6 +85,7 @@
 #![feature(unwind_attributes)]
 #![cfg_attr(stage0, feature(simd))]
 #![cfg_attr(not(stage0), feature(repr_simd, platform_intrinsics))]
+#![feature(slice_patterns)]
```
Comment (alexcrichton) [inline src/libcore/lib.rs]: We've historically had a lot of codegen issues with this in the past, so I'm somewhat wary to enable this to implement this for now. Can this be implemented without this feature?

**Comment by tbu-** [comment]
Making empty array `Copy` requires a compiler change because the compiler will give you an error if you try to implement `Copy` for an array.

This was possible to implement without using slice patterns, but only in a weird way.

r? @alexcrichton

File: src/libcore/array.rs [hunk 2]
```diff
@@ -210,3 +203,31 @@ macro_rules! array_impl_default {
...
 
 array_impl_default!{32, T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T}
+
+macro_rules! array_impl_clone {
+    {$n:expr, $i:expr, $($idx:expr,)*} => {
+        #[stable(feature = "rust1", since = "1.0.0")]
+        impl<T: Clone> Clone for [T; $n] {
+            fn clone(&self) -> [T; $n] {
+                let temp = [&self[$i], $(&self[$idx]),*];
+                [temp[$i].clone(), $(temp[$idx].clone()),*]
```
Comment (eefriedman) [inline src/libcore/array.rs]: I think you can make this a bit simpler:

```
            fn clone(&self) -> [T; $n] {
                [temp[$i-$i].clone(), $(temp[$i-$idx].clone()),*] 
            }
```

Comment (tbu-) [inline src/libcore/array.rs]: Good idea, thank you. For some reason I couldn't figure out how to reverse the indices.

**Comment by bluss** [comment]
It's a breaking change, but only for relatively broken code. It's simply that copyable elements have .clone() called now too, so the cloning is now user-observable, and is a user-observable change.

I think we can say that any code that breaks is already buggy (due to clone and copy not being equivalent).

Example:

```rust
#[derive(Copy)]
struct Foo;

impl Clone for Foo { fn clone(&self) -> Self { panic!() } }

// Will panic in new code, used to be “fine.”
[Foo].clone(); 
```

**Comment by alexcrichton** [comment]
Hm upon reflection I'm a little less cure about special casing a 0-length array, for example do we special case that for `Send` and `Sync` as well (e.g. other marker traits), I'd want to make sure we stay consistent there.

Also yeah technically this could break code, but I can't imagine anyone actually relying on that in practice. You're likely to have way more surprising results than this if your `Clone` is that different than `Copy`.

**Comment by tbu-** [comment]
@alexcrichton Should I split that out into a different PR again?

**Comment by alexcrichton** [comment]
Hm yeah, sorry for the roundabout there!

**Comment by tbu-** [comment]
@alexcrichton Split it up again.

**Comment by netvl** [comment]
I'm not sure why, but it seems that this PR broke compilation of structures like
```rust
#[derive(Copy, Clone)]
struct Array {
    arr: [[u8; 256]; 4]
}
```
that is, structures with `Copy, Clone` derive attribute which contain a short array of longer arrays. Somehow, previously this code did compile. One of affected libraries seems to be rust-crypto (DaGenix/rust-crypto#335).

**Comment by bluss** [comment]
@netvl That's a clear case that we didn't think about :(  The reason is that `[_; 256]` does not implement `Clone`.

**Pull request merged**